src/activation.rs

```diff
@@ -0,0 +1,158 @@
+use wayland_client::{
+    globals::{BindError, GlobalList},
+    protocol::{wl_seat, wl_surface},
+    Dispatch, Proxy, QueueHandle,
+};
+use wayland_protocols::xdg::activation::v1::client::{xdg_activation_token_v1, xdg_activation_v1};
+
+use crate::{
+    error::GlobalError,
+    globals::{GlobalData, ProvidesBoundGlobal},
+};
+
+/// Data attached to a token request
+pub trait RequestData: Send + Sync {
+    /// App_id of the application requesting the token, if applicable
+    fn app_id(&self) -> Option<String>;
+    /// Seat and serial of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests without
+    /// recent serials. There is no way to detect this from the client-side.
+    fn serial(&self) -> Option<(&wl_seat::WlSeat, u32)>;
+    /// Surface of the window requesting the token, if applicable.
+    ///
+    /// *Warning*: Many compositors will issue invalid tokens for requests from
+    /// unfocused surfaces. There is no way to detect this from the client-side.
+    fn surface(&self) -> Option<&wl_surface::WlSurface>;
+}
+
+impl RequestData
+    for (Option<(wl_seat::WlSeat, u32)>, Option<wl_surface::WlSurface>, Option<String>)
+{
+    fn serial(&self) -> Option<(&wl_seat::WlSeat, u32)> {
+        self.0.as_ref().map(|(seat, serial)| (seat, *serial))
+    }
+
+    fn surface(&self) -> Option<&wl_surface::WlSurface> {
+        self.1.as_ref()
+    }
+
+    fn app_id(&self) -> Option<String> {
+        self.2.clone()
+    }
```

Comment:
User can clone if they want to.

```suggestion
    fn app_id(&self) -> Option<&str> {
        self.2.as_ref()
    }
```

Reply:
Yes, it is public api, but the only consumer is likely the xdg-activation implementation. Anyway changing this.